Fix build errors on MSVC

chrisps opened the pull request:
This wraps some gnu specific things so that people can build on MSVC. I wanted to test this interpreter out, but had to make these changes first.

disjtqz committed:
MSVC compat, wrap attr visibility and noinline in macros

disjtqz committed:
Use variadic macro params instead to wrap debug checks to fix err on msvc

disjtqz committed:
Redirect Mmap functions to virtalloc on windows to fix msvc build error

disjtqz committed:
Fix linker error by stubbing sys_getallocatedblocks_impl

kmod commented on Python/ceval.c:
```diff
@@ -5012,8 +5024,12 @@ call_function_ceval(PyThreadState *tstate, PyObject ***pp_stack, Py_ssize_t opar
+#ifndef _MSC_VER
     if (__builtin_expect(tstate->use_tracing, 0)) {
+#else
+     if ((tstate->use_tracing, 0)) {
```
I think you need to remove the `0` (and parenthesis) here. Also the whitespace seems a little bit funky

kmod commented on Include/ceval.h:
```diff
@@ -87,8 +87,8 @@ PyAPI_FUNC(void) Py_SetRecursionLimit(int);
-#define Py_EnterRecursiveCall(where) (0)
-#define Py_LeaveRecursiveCall(where) ((void)0)
+#define Py_EnterRecursiveCall(...) (0)
+#define Py_LeaveRecursiveCall(..) ((void)0)
```
I'm curious, why is this needed? Also are there supposed to be 3 dots and then 2 dots?

chrisps replied:
Errors out on wrong number of arguments to the macro on MSVC.
The two dots is my mistake

kmod replied:
What I meant was "why is a difference from CPython needed here", and it looks like the answer is because I messed up the conversion of Py_LeaveRecursiveCall. I think the proper solution here is to keep the definition of Py_EnterRecursiveCall and to remove the erroneous argument from Py_LeaveRecursiveCall

kmod commented on Objects/obmalloc.c:
```diff
@@ -145,7 +145,16 @@ _PyObject_ArenaVirtualFree(void *ctx, void *ptr, size_t size)
 {
     VirtualFree(ptr, 0, MEM_RELEASE);
 }
+static void *
```
I see what you're doing here but I think a better fix might be to change the callers of _PyObject_ArenaMmap to call the correct function

chrisps replied:
I think just having a single wrapper around virtual memory allocation/deallocation is more convenient and makes more sense overall.

kmod replied:
Sure it could be a function wrapper or a macro wrapper, but I think it's incorrect to say that Mmap is implemented using VirtualAlloc.

kmod commented on Python/sysmodule.c:
```diff
@@ -1732,9 +1732,13 @@ static Py_ssize_t
 sys_getallocatedblocks_impl(PyObject *module)
 /*[clinic end generated code: output=f0c4e873f0b6dcf7 input=dab13ee346a0673e]*/
 {
+    #if PY_DEBUGGING_FEATURES
```
I'm curious, what's the link error here?

chrisps replied:
sys_getallocatedblocks_impl gets referenced later despite being ifdefed out. I stubbed to remove the error instead of removing the reference since its a documented member of sys

kmod replied:
I'm confused about what's going on here, is MSVC trying to compile ifdef'd out code? What's the error you get?

I removed the function because I believe it's better to not have the function than to have it behave incorrectly. It looks like the removal was unnecessary, so if you want you could just remove the `PY_DEBUGGING_FEATURES` ifdefs.

kmod reviewed (changes requested):
Thanks so much for looking into this! I left some comments on your changes.

I think the bigger comment is that, sadly, windows support will be much harder than just fixing the build errors, since our JIT will need to be updated and I'm not sure what else. But this seems like a good first step!

disjtqz committed:
fix accidental ,0

disjtqz committed:
Update ceval.h

chrisps commented:
I moved over just the changes I had that were related to getting it to compile to this repo, I seem to have made two mistakes doing so, those are now corrected.

disjtqz committed:
fix mismatched whitespace
